### PR TITLE
fix: fix nested subcollection initialization

### PR DESCRIPTION
**Pull request by joemckie**
Sub-collections are currently only being initialised one level deep. This PR fixes this to allow subcollections to have subcollections

### CODE REVIEW

**Comment by wovalle** [comment]
I don't think I understand this part: *Sub-collections are currently only being initialised one level deep*. AFAIK you could only have 1 subCollection level. Can subcollections contain other... subcollections?

**Comment by joemckie** [comment]
@wovalle Collections can be infinitely nested, AFAIK.

**Comment by joemckie** [comment]
Screenshot 2019-11-04 at 15 05 54

**Comment by joemckie** [comment]
FYI this also fixes a bug where subcollections are instantiated as their parent collection 😄

**Comment by wovalle** [comment]
I don't know why I had the impression that they could be one level long 🤦🏽‍♂️

File: src/AbstractFirestoreRepository.ts [hunk 1]
```diff
@@ -108,14 +108,15 @@ export abstract class AbstractFirestoreRepository<T extends IEntity>
     if (!doc.exists) {
       return null;
     }
+    
     // tslint:disable-next-line:no-unnecessary-type-assertion
     const entity = plainToClass(
-      this.colMetadata.entity as any,
+      this.colMetadata.entity,
```
Comment (wovalle) [inline src/AbstractFirestoreRepository.ts]: Interested in why suddendly this `any` is not required, I was getting compiler errors before 🤔

File: src/AbstractFirestoreRepository.ts [hunk 2]
```diff
@@ -108,14 +108,15 @@ export abstract class AbstractFirestoreRepository<T extends IEntity>
...
       this.transformFirestoreTypes(doc.data() as T)
     ) as T;
-    if (this.collectionType === FirestoreCollectionType.collection) {
```
Comment (wovalle) [inline src/AbstractFirestoreRepository.ts]: If this check is not needed anymore, shouldn't we remove the `collectionType` field altogether?

**Comment by wovalle** [comment]
I don't know why I had the impression that they could be one level long 🤦🏽‍♂️

**Comment by joemckie** [comment]
Easy mistake to make, I think 😄

I realised this is actually quite a bit more work than originally anticipated... After trialling it in my project, nested subcollections are saved at root level under their parent collection, so an `AlbumImage` would be saved like so:

- Bands
- - Albums
- Albums
- - AlbumImage

This is basically because the repository assigns the `firestoreColRef` only based on the `colName` and `subColName`. I've been attempting a way of building the document path based on the collection's `parentEntity`, which should allow for a proper firestore path to be used. The only issue with this is if a sub collection is used by multiple collections, we would need a way to correctly assign the grandparent entity when assigning a subcollection. 🤔

**Comment by joemckie** [comment]
Also we would need some way of finding the correct `docId` all the way to the root...

**Comment by joemckie** [comment]
@wovalle okay think I have it! It required a rework of how repositories work. They are now initialised with a document path, and then when subcollections are created, it creates a new repository with the document id and subcollection name appended to the document path (this is how it can recurse infinitely)

**Comment by wovalle** [comment]
Instead of initializing with colName/subColId/docId passing a `path` as param? I'm quite intrigued with the change that it'll require but I do undestand that's the way to go.

Initially fireorm was built with the assumption that firestore only accepted one nested SubCollection but that was wrong. Initializing with colname/subcolname quickly became cumbersome (checking if `docId` is defined to know if you're in a subcollection, that's not flexible enough). I understand that the initial implementation doesn't scale so go ahead and update it.

I'll keep an eye to this PR because I know monumental changes are coming 😂

**Comment by joemckie** [comment]
@wovalle thank you! This should be a lot more flexible overall I think 😄

**Comment by joemckie** [comment]
@wovalle okay I've made the changes, I'll give it a trial run in a live project tonight to see if it works 😉

File: src/AbstractFirestoreRepository.ts [hunk 3]
```diff
@@ -25,43 +24,37 @@ import { ValidationError } from 'class-validator';
...
+  protected readonly subColMetadata: CollectionMetadata[];
   protected readonly colMetadata: CollectionMetadata;
   protected readonly collectionType: FirestoreCollectionType;
   protected readonly colName: string;
-  protected readonly docId: string;
-  protected readonly subColName: string;
   protected readonly config: MetadataStorageConfig;
 
+  // TODO: Make this private when transactions have been fixed
+  public readonly collectionPath: string;
```
Comment (joemckie) [inline src/AbstractFirestoreRepository.ts]: I've made this public, as having the collection path can be a helpful workaround whilst transactions are being fixed :)

**Comment by joemckie** [comment]
Just tested this out in a project, works like a charm 👌

**Comment by wovalle** [comment]
@joemckie can you suggest an scenario so I can create an integration this when I finish with the transaction refactor?

Just adding a collection that contains a subcollection that contains another subcollection and test something in the repository of all the collections can make it?

**Comment by joemckie** [comment]
@wovalle exactly, that should work well enough for an integration test - no need to go too deep

**Comment by manwithsteelnerves** [comment]
@wovalle Can you please let me know the status of nested subcollections more than one level?

It would be great to see an example as currently I see the first subcollection won’t have  any class level decoratir so wondering how to declare a subcollection with in that class.

Also, nested subcollections work for transactions too?

**Comment by joemckie** [comment]
@manwithsteelnerves I was the one that worked on this and have been using the functionality in a project of my own for a while - it works as expected 😄

You can create sub-collections in the same fashion as a usual collection, (just use `@Collection`), then just assign the class to a property in the parent collection like in [here](https://fireorm.js.org/#/Subcollections).

I think transactions are currently being worked on by @wovalle so they don't work just yet.

Hope this helps

**Comment by joemckie** [comment]
I think the docs may need an update to include the Collection decorator when creating subcollections 😄

**Comment by manwithsteelnerves** [comment]
@joemckie I saw your commits but they don’t seem to be in the documentation yet. As still doc points at using subcollection decorator. I liked the point mentioned that subcollections are collections with a parent.

Regarding transactions, can i get the code related to global transaction suggested by @wovalle ? Or any alternative?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by joemckie** [comment]
> @joemckie I saw your commits but they don’t seem to be in the documentation yet. As still doc points at using subcollection decorator. I liked the point mentioned that subcollections are just collections with a parent, and no different.

Yeah, think the docs just need a quick PR to help explain things better. I probably missed some parts.

> Lastly, i see fetching the subcollection needs a read of the document which is kind of bothering me. Any workaround for this?

You can provide a path name to `getRepository`, which is a method I have been using to access deeply nested collections. Given an entity `Product` that is a subcollection of `Shop`, you can use:

 ```ts
getRepository(Product, 'Shops/:shopId/products')
```

**Comment by manwithsteelnerves** [comment]
> getRepository(Product, 'Shops/:shopId/products')
Thats great! I need to give a try. Too much to expect but, Any way to make it type safe? As it avoids lots of errors. Usually I like having an error for any variable name change as its easy to maintain. Else lot of tests need to be done for a single variable name change.

**Comment by wovalle** [comment]
Thanks for answering @joemckie! I didn't even know the `collectionPath` trick 😅

**Comment by joemckie** [comment]
> Any way to make it type safe? As it avoids lots of errors. Usually I like having an error for any variable name change as its easy to maintain.

I'm not entirely sure the collection path could be type-safe as it would require knowledge of all the collections and their paths beforehand. I don't even think a helper function to build the path from an entity would work without grabbing the parent collection first, as there is no path provided to it. A workaround might be to provide the pluralised version of the entity name (as it would be saved into the database) and then use that in the string, rather than typing it manually? That would avoid some issues that may arise when changing collection names.

> Also, this hidden gem needs a mention in the docs too, unfortunately I didn't find it :)

Will definitely work on that when I get a sec 😄

**Pull request merged**